elixir/apps/domain/lib/domain/clients/client/changeset.ex

```diff
@@ -1,32 +1,131 @@
 defmodule Domain.Clients.Client.Changeset do
   use Domain, :changeset
+  import Ecto.Query
   alias Domain.{Version, Auth, Actors}
   alias Domain.Clients
 
-  @upsert_fields ~w[external_id last_used_token_id name public_key]a
-  @conflict_replace_fields ~w[name
-                              public_key
-                              last_used_token_id
-                              last_seen_user_agent
-                              last_seen_remote_ip
-                              last_seen_remote_ip_location_region
-                              last_seen_remote_ip_location_city
-                              last_seen_remote_ip_location_lat
-                              last_seen_remote_ip_location_lon
-                              last_seen_version
-                              last_seen_at
-                              identity_id
-                              updated_at]a
+  @required_fields ~w[external_id last_used_token_id name public_key]a
+  @hardware_identifiers ~w[device_serial device_uuid identifier_for_vendor firebase_installation_id]a
+  @upsert_fields @required_fields ++ @hardware_identifiers
   @update_fields ~w[name]a
-  @required_fields @upsert_fields
 
   # WireGuard base64-encoded string length
   @key_length 44
 
   def upsert_conflict_target,
     do: {:unsafe_fragment, ~s/(account_id, actor_id, external_id) WHERE deleted_at IS NULL/}
 
-  def upsert_on_conflict, do: {:replace, @conflict_replace_fields}
+  def upsert_on_conflict do
+    # dynamic = dynamic([clients: client], )
+
+    Clients.Client.Query.all()
+    |> update([clients: clients],
+      set: [
+        name: fragment("EXCLUDED.name"),
+        public_key: fragment("EXCLUDED.public_key"),
+        last_used_token_id: fragment("EXCLUDED.last_used_token_id"),
+        last_seen_user_agent: fragment("EXCLUDED.last_seen_user_agent"),
+        last_seen_remote_ip: fragment("EXCLUDED.last_seen_remote_ip"),
+        last_seen_remote_ip_location_region:
+          fragment("EXCLUDED.last_seen_remote_ip_location_region"),
+        last_seen_remote_ip_location_city: fragment("EXCLUDED.last_seen_remote_ip_location_city"),
+        last_seen_remote_ip_location_lat: fragment("EXCLUDED.last_seen_remote_ip_location_lat"),
+        last_seen_remote_ip_location_lon: fragment("EXCLUDED.last_seen_remote_ip_location_lon"),
+        last_seen_version: fragment("EXCLUDED.last_seen_version"),
+        last_seen_at: fragment("EXCLUDED.last_seen_at"),
+        identity_id: fragment("EXCLUDED.identity_id"),
+        device_serial: fragment("EXCLUDED.device_serial"),
+        device_uuid: fragment("EXCLUDED.device_uuid"),
+        identifier_for_vendor: fragment("EXCLUDED.identifier_for_vendor"),
+        firebase_installation_id: fragment("EXCLUDED.firebase_installation_id"),
+        updated_at: fragment("NOW()"),
+        verified_at:
```

Comment:
Yes, this is ugly. I will try to make it better tomorrow but not 100% sure there is way to do it that won't be even worse 🤣